openwrt/wifi-disconnect-low-signal/wifi-disconnect-low-signal

```diff
@@ -3,18 +3,70 @@
 START=99
 
 USE_PROCD=1
-NAME=wifi-disconnect-low-signal.lua
-PROG=/usr/bin/wifi-disconnect-low-signal.lua
+PACKAGE=wireless
+TYPE=wifi-disconnect-low-signal
+NAME="${TYPE}.lua"
+PROG="/usr/bin/${TYPE}.lua"
 
-start_service() {
+VERBOSITY=0
+STDERR=0
+
+validate_wifi_disconnect_low_signal_section() {
+	uci_load_validate "$PACKAGE" "$TYPE" "$1" "$2" \
+		"verbosity:range(0,3):${VERBOSITY}" \
+		"stderr:bool:${STDERR}"
+}
+
+configure_wifi_disconnect_low_signal() {
+	configure_wifi_disconnect_low_signal() {
```

Comment:
what's up in here?

Reply:
got it, but is this a common practice with openwrt/uci? Otherwise I'd skip this, or go with a more standard way that checks instead of redefining it (I was actually confused if it was really intentional or if it was a mistake with patch conflict resolution).

Reply:
btw, don't bother with 2 independent PRs, we're low traffic. It's okay for me to review the other, handling the rssi alongside with this one. (It's also okay for me to review the other PR, I'm ignoring the first commits as they should go away when you rebase, remaining only the rssi work).

Reply:
> [I]s this a common practice with openwrt/uci?

I've only checked the initscripts on my LAN's systems, so I'm not sure I can speak to common practices, but I haven't come across any instances of function redefinition.

As for services that use anonymous sections (`config whatever`, rather than `config whatever 'my-section'`), I've encountered these approaches:

1. Each section is an "instance".  For example, each `config dropbear` section in the Dropbear configuration file represents the configuration for a single `dropbear` daemon, and specifying multiple sections results in running multiple daemons.
2. Each section's settings clobber the settings from previous sections.  This is how `/etc/init.d/system` works: only the last `config system` section's settings have any effect.
3. The service eschews `uci_load_validate` and instead shells out to `uci` directly.  For instance, `/etc/init.d/rngd`, which uses `uci get` to pull settings from `system.@rngd[0]`, i.e. the first-defined `config rngd` entry in `/etc/config.system`:

```sh
#!/bin/sh /etc/rc.common
# Copyright (C) 2011-2014 OpenWrt.org

START=25

USE_PROCD=1
PROG=/sbin/rngd

start_service() {
        local enabled=$(uci -q get system.@rngd[0].enabled)
        local precmd=$(uci -q get system.@rngd[0].precmd)
        local device=$(uci -q get system.@rngd[0].device)
        local watermark=$(uci -q get system.@rngd[0].fill_watermark)

        [ "$enabled" = "1" ] || return

        [ -z "$precmd" ] || ${precmd} ${device}

        [ -z "$device" ] || device="-r ${device}"
        [ -z "$watermark" ] || watermark="-W ${watermark}"

        procd_open_instance
        procd_set_param command "$PROG" -f ${device} ${watermark}
        procd_set_param stderr 1
        procd_close_instance
}
```

The first approach is not relevant here, since (AFAIK?) only one instance of `wifi-disconnect-low-signal.lua` should be running at any given time.  The third approach seems fine, except it lacks `uci_load_validate`'s convenient input validation.  And the second approach is liable to produce surprising results; for instance, take the following example `/etc/config/system`:

```
config system
    option hostname 'mydevice'
    option zonename 'America/New_York'

config system
    option log_size 128
```

A system with this configuration will end up with the hostname `OpenWrt` and the zonename `UTC`, because these are the default values for the `hostname` and `zonename` options, and when `/etc/init.d/system` applies the settings from the second `config system` section, those default clobber the custom settings from the first `config system` section.  This is not the result I would have expected before peering into the initscript, and it's all the worse because the clobbering is silent: `/etc/init.d/system` does not notify about multiple `config system` sections.  Note that the same is true of `/etc/init.d/rngd`.

> [I'd] go with a more standard way that checks instead of redefining it[.]

I poked around in `/lib/functions.sh` and friends, and as far as I can tell there is no public interface for determining the number of sections of a given type. [This OpenWrt Forum thread](https://forum.openwrt.org/t/number-of-sections-in-a-config-file/47132) has a few suggestions for how to count config sections, but in my estimation they are all a bit gross.

In any case, I haven't found an initscript that tries to detect the number of sections.  I can yank the detection code from `/etc/config/wifi-disconnect-low-signal`, but this comes with the downsides mentioned above.

Or I may be able to put something together with [`config_cb`](https://openwrt.org/docs/guide-developer/config-scripting#config_cb_callback) and/or [`option_cb`](https://openwrt.org/docs/guide-developer/config-scripting#option_cb_callback) that gracefully merges options from multiple sections without the clobbering-by-defaults problem mentioned earlier.  This approach may require hand-rolled input validation, however.

Would you accept a comment above the redefined `configure_wifi_disconnect_low_signal` explaining its purpose?

Alternately, I could replace the function redefinition with a global variable counting the number of times `configure_wifi_disconnect_low_signal` has been called, issuing a warning when the variable is > 1.

> btw, don't bother with 2 independent PRs, we're low traffic. It's okay for me to review the other, handling the rssi alongside with this one. (It's also okay for me to review the other PR, I'm ignoring the first commits as they should go away when you rebase, remaining only the rssi work).

Which do you prefer?  I'd be happy to close #22 and include its changes in this PR.

Reply:
Aaaand after all that, I think I've found a better way: just don't provide default option values when calling `uci_load_validate`, and only set `VERBOSITY` and  `STDERR` if `verbosity` and `stderr` are nonempty.  Interested to hear what you thing of this approach.